src/coreclr/jit/optimizer.cpp

```diff
@@ -7067,53 +7153,113 @@ bool Compiler::optVNIsLoopInvariant(ValueNum vn, unsigned lnum, VNToBoolMap* loo
     return res;
 }
 
-/*****************************************************************************
- *
- *  Creates a pre-header block for the given loop - a preheader is a BBJ_NONE
- *  header. The pre-header will replace the current lpHead in the loop table.
- *  The loop has to be a do-while loop. Thus, all blocks dominated by lpHead
- *  will also be dominated by the loop-top, lpHead->bbNext.
- *
- */
-
+//------------------------------------------------------------------------------
+// fgCreateLoopPreHeader: Creates a pre-header block for the given loop.
+// A pre-header is a block outside the loop that falls through or branches to the loop
+// entry block. It is the only non-loop predecessor block to the entry block (thus, it
+// dominates the entry block). The pre-header replaces the current lpHead in the loop table.
+// The pre-header will be placed immediately before the loop top block, which is the first
+// block of the loop in program order.
+//
+// Once a loop has a pre-header, calling this function will immediately return without
+// creating another.
+//
+// If there already exists a block that meets the pre-header requirements, that block is marked
+// as a pre-header, and no flow graph modification is made.
+//
+// Note that the pre-header block can be in a different EH region from blocks in the loop, including the
+// entry block. Code doing hoisting is required to check the EH legality of hoisting to the pre-header
+// before doing so.
+//
+// Since the flow graph has changed, if needed, fgUpdateChangedFlowGraph() should be called after this
+// to update the block numbers, reachability, and dominators. The loop table does not need to be rebuilt.
+// The new pre-header block does have a copy of the previous 'head' reachability set, but the pre-header
+// itself doesn't exist in any reachability/dominator sets. `fgDominate` has code to specifically
+// handle queries about the pre-header dominating other blocks, even without re-computing dominators.
+// The preds lists have been maintained.
+//
+// Currently, if you create a pre-header but don't put any code in it, any subsequent fgUpdateFlowGraph()
+// pass might choose to compact the empty pre-header with a predecessor block. That is, a pre-header
+// block might disappear if not used.
```

Comment:
I think we should reconsider this compaction -- those edges tend to be critical edges and we're probably better off having blocks there, at least for most of the phases.

Eg it might help prevent LSRA from putting its resolution move blocks in strange places.

We'd probably need a late flow opts phase that then does the cond->(empty) uncond opt.

Reply:
It might not be hard to prevent compaction or cond->(empty)uncond opt. when a pre-header is involved, but as you point out, we might need a late flow opts. phase that we currently don't have to clean these up. So put it on the list for future work?

Reply:
Yes. We should consider deferring these simple flow graph updates until some (new) late phase.